src/detail/fastrtps_dynamic_type.cpp

```diff
@@ -35,105 +34,122 @@
   const rosidl_dynamic_typesupport_dynamic_type_impl_t * other_type_impl,
   bool * equals)
 {
-  (void) serialization_support_impl;
-  auto type = eprosima::fastrtps::types::DynamicType_ptr(
-    *static_cast<const eprosima::fastrtps::types::DynamicType_ptr *>(type_impl->handle));
-  auto other = eprosima::fastrtps::types::DynamicType_ptr(
-    *static_cast<const eprosima::fastrtps::types::DynamicType_ptr *>(other_type_impl->handle));
+  (void)serialization_support_impl;
+  auto type =
+    static_cast<const fastdds__rosidl_dynamic_typesupport_dynamic_type_impl *>(type_impl->handle);
+  auto other =
+    static_cast<const fastdds__rosidl_dynamic_typesupport_dynamic_type_impl *>(other_type_impl->
+    handle);
 
-  *equals = type->equals(other.get());
+  *equals = type->ref_type->equals(other->ref_type);
   return RCUTILS_RET_OK;
 }
 
-
-rcutils_ret_t
-fastrtps__dynamic_type_get_member_count(
+rcutils_ret_t fastdds__dynamic_type_get_member_count(
   rosidl_dynamic_typesupport_serialization_support_impl_t * serialization_support_impl,
   const rosidl_dynamic_typesupport_dynamic_type_impl_t * type_impl,
   size_t * member_count)
 {
-  (void) serialization_support_impl;
-  auto type = eprosima::fastrtps::types::DynamicType_ptr(
-    *static_cast<const eprosima::fastrtps::types::DynamicType_ptr *>(type_impl->handle));
+  (void)serialization_support_impl;
+  auto type =
+    static_cast<const fastdds__rosidl_dynamic_typesupport_dynamic_type_impl *>(type_impl->handle);
 
-  *member_count = type->get_members_count();
+  *member_count = type->ref_type->get_member_count();
   return RCUTILS_RET_OK;
 }
 
-
 // DYNAMIC TYPE CONSTRUCTION =======================================================================
 rcutils_ret_t
-fastrtps__dynamic_type_builder_init(
+fastdds__dynamic_type_builder_init(
   rosidl_dynamic_typesupport_serialization_support_impl_t * serialization_support_impl,
-  const char * name, size_t name_length,
+  const char * name,
+  size_t name_length,
   rcutils_allocator_t * allocator,
   rosidl_dynamic_typesupport_dynamic_type_builder_impl_t * type_builder_impl)
 {
-  (void) allocator;
+  (void)allocator;
+
+  auto fastdds_impl =
+    static_cast<fastdds__serialization_support_impl_handle_t *>(serialization_support_impl->handle);
 
-  auto fastrtps_impl = static_cast<fastrtps__serialization_support_impl_handle_t *>(
-    serialization_support_impl->handle);
-  DynamicTypeBuilder * type_builder_handle = fastrtps_impl->type_factory_->create_struct_builder();
+  // We must replace "/" with "::" in type names
+  std::string name_string = fastdds__replace_string(std::string(name, name_length), "/", "::");
+
+  auto type_builder_handle =
+    new (std::nothrow) fastdds__rosidl_dynamic_typesupport_dynamic_type_builder_impl();
   if (!type_builder_handle) {
     RCUTILS_SET_ERROR_MSG("Could not init new struct type builder");
-    return RCUTILS_RET_BAD_ALLOC;
+    return RCUTILS_RET_ERROR;
```

Comment:
I think that `RCUTILS_RET_BAD_ALLOC` is correct here:

```suggestion
    return RCUTILS_RET_BAD_ALLOC;
```

Reply:
Done in 2ec21b3